### ISSUE TITLE
Schedules Broadcasting

### ISSUE THREAD

**Issue by scienceworld** [museumvictoria/nodel#21]
Seems like we have a weird case where schedules seem to be affecting Exhibit nodes that are not associated to a gallery or a schedule or associated to a different schedule.

Issue 1.  We had some galleries scheduled with exceptions to not turn off at 6pm and to turn off at 11pm.  The gallery turned off at 6pm.

I can see the weekend scheduled for AllOff had an exception for the day in question and then a corresponding new schedule to turn it off at 11.

Issue 2.  I added a new exhibit and didnt associate it to a gallery or schedule for its enable and disable events.  The exhibit powered up this morning at 9 and turned off at 5 which is the normal gallery on/off time for the rest of the building.

![image](https://cloud.githubusercontent.com/assets/3997916/5641821/459d16be-95f4-11e4-84aa-b5e7471ebda9.png)

Its almost like all exhibit nodes are listening to the schedule even if the exhibit isnt associated to it.

**Comment by justparking** [comment]
It should be highly unlikely a Node is able to recieve stray events from another node it has no bindings to. Can you see if it's possible to reproduce this on demand if you manually trigger the events?

Regardless, if you enable program logging and pass that on, I should be able to follow this up. Cheers.

**Comment by scienceworld** [comment]
Issue 2:
I think it was a stale binding issue.
I updated nodel to 2.05 yesterday and had to re-launch nodel on our host.  all the bindings were re-learned and the exhibit did not receive a rogue on command today.

One of the things I do when setting up a new node is copying an existing node over, including its config.  then I update the bindings to reflects its new role.  so this exhibit would have been bound to a different gallery.

logs from host.
http://we.tl/jvp9eHaoQ4 (link valid for 7 days)

node in question: "OmnimaxXiboExhibit"

the node did not receive a scheduled on command today, which is what is desired.


Issue 1:
this could also be a stale binding issue as most of the nodes were added by copying them from somewhere else.

eg node that was shutdown was "wheelchairracerexhibit"
which is bound to "2ndcirculationgallery" and controlled by "2ndcircuationschedule"
2nd circuation schedule had an exclusion on the saturday shutdown for 1/3/2015
there was a scheduled off command for 23:00 on 1/3/2015.
schedule node here: http://we.tl/qm12d0RO2b (link valid for 7 days)
here are the host logs from jan 4th, which should include everything that happened on the 3rd when other galleries shut down that weren't scheduled to. http://we.tl/3nvTvCasF0

**Comment by scienceworld** [comment]
I think I have definitively figured out the issue.

If you edit an Exhibit's event disable or enable bindings and change the binding From GalleryA to GalleryB the node stays bound to GalleryA and is also bound to GalleryB events.

It seems that nodel doesn't delete a binding on a field change, but does add the new binding on the field change. 

Re-launching the nodel service on the host will cause all new bindings to be re-learned.  So the issue goes away at that point.

So in Summary
Confirmed that if you edit an exhibit and change the node that issues the Enable and Disable signals the exhibit will listen to both the old and new node until the nodel service is restarted on the host.

**Comment by justparking** [comment]
That's great you've been able to reproduce it. I'll attempt to do the same here...

**Comment by scienceworld** [comment]
yes, it's reproducible

thank you.

**Comment by justparking** [comment]
Thanks for spotting the problem... I've released v2.0.6. It was caused by some overzealous code refactoring that occurred when Nodel was converted from its original repository and published on this github one. Two private fields were being "overshadowed" by their subclassed ones; the side-effect being part of the cleanup routine would fail.

I removed those duplicate fields and the clean up routines now work the way they originally did. We must've have seen this recently because most of the systems have been stable for a while now.

It's possible some of other recent issues related to events (duplication, being missed, etc) may be related to this bugfix. If you're able to reproduce those easily, and find they now work, we can close them off too.
